apps/engine-test/src/orbitControls.js

```diff
@@ -0,0 +1,235 @@
+/** Based on orbitControls from @openjscad/regl-renderer 
+ and https://webglfundamentals.org/webgl/lessons/webgl-3d-camera.html
+ and
+   https://elalish.blogspot.com/2022/04/3d-interaction.html
+   https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/three-components/SmoothControls.ts
+*/
+import * as mat4 from 'gl-matrix/esm/mat4.js'
+import * as vec3 from 'gl-matrix/esm/vec3.js'
+
+const { hypot, acos, asin, PI, cos, sin } = Math
+const PId2 = PI / 2
+const TAU = PI * 2
+
+export class OrbitCamera {
+  position = [180, -180, 220]
+  target = [0, 0, 0]
+}
+
+export class OrbitState {
+  position
+  target
+  calc
+  constructor(position = [180, -180, 220], target = [0, 0, 0], ...args) {
+    if (args.length) {
+      // clone
+      this.position = position
+      this.target = target
+      this.calc = args[0]
+    } else {
+      this.set(position, target, false)
+    }
+  }
+
+  set(position, target, fire = true) {
+    if (target) this.target = target
+    if (position) this.position = position
+    this.updateCalc()
+    if (fire) this.fireChange()
+  }
+
+  updateCalc() {
+    this.calc = camRotation({ target: this.target, position: this.position })
+  }
+
+  setRotate(rx = 0, rz = 0, target) {
+    if(target) this.target = target
+    this.calc.rx = rx
+    this.calc.rz = rz
+    this.position = calcCamPos({ ...this.calc, target: this.target })
+    this.fireChange()
+  }
+
+  rotateBy(rx = 0, rz = 0) {
+    this.calc.rx += rx
+    if (this.calc.rx < 0) this.calc.rx = 1.0e-10
+    if (this.calc.rx > PI) this.calc.rx = PI
+    this.calc.rz += rz
+    this.position = calcCamPos({ ...this.calc, target: this.target })
+    this.fireChange()
+  }
+
+  panBy(dx, dy) {
+    const { rx, rz } = this.calc
+    let pan = [dx, -dy, 0]
+    pan = vec3.transformMat4([], pan, fromXZRotation(rx, rz))
+    this.moveBy(pan)
+  }
+
+  moveBy(vec) {
+    this.position = vec3.add([], this.position, vec)
+    this.target = vec3.add([], this.target, vec)
+    this.fireChange()
+  }
+
+  zoomBy(amount) {
+    this.calc.len *= 1 + amount
+    this.position = calcCamPos({ ...this.calc, target: this.target })
+    this.fireChange()
+  }
+
+  clone() {
```

Comment:
Why a clone fn?

Reply:
it is for future use with animations, OrbitCalc is meant as kind of orbit state, and when I want to animate it I will clone it to have a reference point, then use it to produce intermediate state for animation to next location based on animation percent
.